### PR TITLE
Implement scenes tiles in TileMaps

### PR DESCRIPTION
**Pull request by groud**
Peek 18-05-2021 13-28

Implements a TileSetScenesCollectionSource for TileSets. Each tile in this TileSet source has a scene that will be instantiated as an internal child of the TileMap node.

I also implemented a "display placeholder" property that allow displaying a placeholder at the position of the tile in case a scene is not visible.

Also, I moved `TileSetAtlasSource::INVALID_*` to `TileSetSource::INVALID_*`

### CODE REVIEW

File: scene/resources/tile_set.cpp [hunk 1]
```diff
@@ -2034,6 +2035,203 @@ void TileSetAtlasSource::_compute_next_alternative_id(const Vector2i p_atlas_coo
...
+	return scenes.has(p_alternative_tile);
+}
+
+int TileSetScenesCollectionSource::create_scene_tile(Ref<PackedScene> p_packed_scene, int p_id_override) {
+	ERR_FAIL_COND_V_MSG(p_id_override >= 0 && scenes.has(p_id_override), -1, vformat("Cannot create scene tile. Another scene tile exists with id %d.", p_id_override));
+
+	int new_scene_id = p_id_override >= 0 ? p_id_override : next_scene_id;
+
+	scenes[new_scene_id] = SceneData();
+	;
```
Comment (Calinou) [inline scene/resources/tile_set.cpp:2081]: There is a lone semicolon:

```suggestion
```

File: doc/classes/TileSetScenesCollectionSource.xml [hunk 2]
```diff
@@ -0,0 +1,155 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="TileSetScenesCollectionSource" inherits="TileSetSource" version="4.0">
+	<brief_description>
```
Comment (Calinou) [inline doc/classes/TileSetScenesCollectionSource.xml:3]: Remember to add documentation for this class :slightly_smiling_face:

File: editor/plugins/tiles/tile_map_editor.cpp [hunk 3]
```diff
@@ -1746,6 +1890,28 @@ TileMapEditorTilesPlugin::TileMapEditorTilesPlugin() {
...
 
+	// Scenes collection source.
+	scene_tiles_list = memnew(ItemList);
+	scene_tiles_list->set_h_size_flags(SIZE_EXPAND_FILL);
+	scene_tiles_list->set_v_size_flags(SIZE_EXPAND_FILL);
+	scene_tiles_list->set_drag_forwarding(this);
+	scene_tiles_list->set_select_mode(ItemList::SELECT_MULTI);
+	scene_tiles_list->connect("multi_selected", callable_mp(this, &TileMapEditorTilesPlugin::_scenes_list_multi_selected));
+	scene_tiles_list->connect("nothing_selected", callable_mp(this, &TileMapEditorTilesPlugin::_scenes_list_nothing_selected));
+	/*scene_tiles_list->connect("item_activated", callable_mp(this, &TileSetScenesCollectionSourceEditor::_scenes_list_item_activated));*/
```
Comment (Calinou) [inline editor/plugins/tiles/tile_map_editor.cpp:1901]: Remove commented out code (unless you meant to uncomment it):

```suggestion
```

File: editor/plugins/tiles/tile_set_scenes_collection_source_editor.cpp [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+		return;
+	}
+
+	// Remove listener for old objects.
+	if (tile_set_scenes_collection_source) {
+		tile_set_scenes_collection_source->disconnect("changed", callable_mp(this, &TileSetScenesCollectionSourceEditor::_tile_set_scenes_collection_source_changed));
+	}
+
+	// Clear the selection.
+	// selection.clear();
```
Comment (Calinou) [inline editor/plugins/tiles/tile_set_scenes_collection_source_editor.cpp:350]: Remove commented out code unless you intend to uncomment it:

```suggestion
```

File: scene/resources/tile_set.h [hunk 5]
```diff
@@ -636,6 +682,18 @@ class TileSetAtlasPluginNavigation : public TileSetPlugin {
 	virtual void draw_quadrant_debug(TileMap *p_tile_map, TileMapQuadrant *p_quadrant) override;
 };
 
+class TileSetPluginScenesCollections : public TileSetPlugin {
+	GDCLASS(TileSetPluginScenesCollections, TileSetPlugin);
+
+public:
+	// Tilemap updates
+	//virtual void tilemap_notification(TileMap *p_tile_map, int p_what) override;
```
Comment (Calinou) [inline scene/resources/tile_set.h:690]: Commented out code:

```suggestion
```

File: scene/resources/tile_set.h [hunk 6]
```diff
@@ -636,6 +682,18 @@ class TileSetAtlasPluginNavigation : public TileSetPlugin {
...
+	virtual void update_dirty_quadrants(TileMap *p_tile_map, SelfList<TileMapQuadrant>::List &r_dirty_quadrant_list) override;
+	//virtual void create_quadrant(TileMap *p_tile_map, TileMapQuadrant *p_quadrant) override;
```
Comment (Calinou) [inline scene/resources/tile_set.h:692]: Commented out code:

```suggestion
```

**Review by Calinou — DISMISSED** [review]

File: doc/classes/TileSetScenesCollectionSource.xml (hunk 2 above)
Comment (groud) [reply doc/classes/TileSetScenesCollectionSource.xml:3]: Yeah, I'll do a documentation pass once I'm sure the API is in its final form. For now I only synchronize the docs.

**Comment by KoBeWi** [comment]
- Not sure how you should prevent that, but it's possible to draw the scene that you are currently editing, resulting in cyclic dependency. Although it just won't save, so maybe it's handled correctly.
- I managed to make the editor crash once, not sure how:
```
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] Ref<SceneState>::operator-> (C:\godot_source\core\object\reference.h:109)
[1] PackedScene::instance (C:\godot_source\scene\resources\packed_scene.cpp:1626)
[2] TileSetPluginScenesCollections::update_dirty_quadrants (C:\godot_source\scene\resources\tile_set.cpp:4484)
[3] TileMap::update_dirty_quadrants (C:\godot_source\scene\2d\tile_map.cpp:373)
[4] call_with_variant_args_helper<TileMap> (C:\godot_source\core\variant\binder_common.h:214)
[5] call_with_variant_args_dv<TileMap> (C:\godot_source\core\variant\binder_common.h:357)
[6] MethodBindT<TileMap>::call (C:\godot_source\core\object\method_bind.h:285)
[7] Object::call (C:\godot_source\core\object\object.cpp:784)
[8] Callable::call (C:\godot_source\core\variant\callable.cpp:53)
[9] MessageQueue::_call_function (C:\godot_source\core\object\message_queue.cpp:259)
[10] MessageQueue::flush (C:\godot_source\core\object\message_queue.cpp:306)
[11] SceneTree::physics_process (C:\godot_source\scene\main\scene_tree.cpp:415)
[12] Main::iteration (C:\godot_source\main\main.cpp:2479)
[13] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:622)
[14] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:163)
[15] _main (C:\godot_source\platform\windows\godot_windows.cpp:185)
[16] main (C:\godot_source\platform\windows\godot_windows.cpp:199)
[17] __scrt_common_main_seh (D:\a01\_work\26\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[18] BaseThreadInitThunk
-- END OF BACKTRACE --
```
- Scene tiles are always centered. They either should have configurable offset or at least use the position of instance's root node.
- I get this bug for some reason:
le Bug

**Comment by groud** [comment]
> Not sure how you should prevent that, but it's possible to draw the scene that you are currently editing, resulting in cyclic dependency. Although it just won't save, so maybe it's handled correctly.

I don't think I can do anything about it. TileSet is a resource, so it can't be strictly associated to the scene it is in. So well... as soon as it does not crashes I believe this is fine, as the engine does not checks for cyclical dependencies in general.

> I managed to make the editor crash once, not sure how:

It happens when trying to paint a scene with an invalid scene. It should be fixed now.

> Scene tiles are always centered. They either should have configurable offset or at least use the position of instance's root node.

Fixed.

> I get this bug for some reason:

Fixed too, I screwed things up when rebasing... ^^

**Comment by groud** [comment]
I have no clue why the Android build fails and how to solve it. That's annoying... :/

**Comment by KoBeWi** [comment]
Seems like it's possible to draw 2 scenes at one when you click between tiles:
Bug2
But only horizontally.

**Comment by groud** [comment]
> Seems like it's possible to draw 2 scenes at one when you click between tiles:

Oh oh, that was a tricky one. My _draw_line function took one Vector2i and one Vector2 as arguments, while they should have been both Vector2. This caused very small differences between the two computed ends of a line, even if, in that case, the end and the beginning of the line should were supposed to have the same value.

Anyway, that's fixed now, thanks!

**Review by akien-mga — APPROVED** [review]
I didn't test, but since @KoBeWi did and code seems fine, should be good to go.

**Pull request merged**